File: python/tradingflow/predictor/variance/single_index.py
from ._base import VariancePredictor, covariance_predictor
from ._common import single_index_covariance


def build(**kwargs) -> VariancePredictor:
    r"""Constructs a single-index factor-model covariance predictor.

    Fits \(r_i(t) = \alpha_i + \beta_i f(t) + \epsilon_i(t)\) stock by stock
    against the equal-weighted cross-sectional mean return as the market proxy,
    and returns \(\Sigma = \sigma_f^2 \beta\beta^T +
    \mathrm{diag}(\sigma_\epsilon^2)\) — the *SI* estimator of Pantaleo et al.
    (2010).
    """
    return covariance_predictor(single_index_covariance, **kwargs)

File: crates/tradingflow/src/operators/predictor/variance/single_index.rs
use super::VariancePredictor;
use crate::operators::predictor::Config;
use crate::python::py_segment_module;

/// Single-index factor model: one market factor plus idiosyncratic variance.
///
/// Regresses each stock on the equal-weighted cross-sectional mean return as a
/// market proxy and returns `Σ = σ_f² ββᵀ + diag(σ_ε²)` — the *SI* estimator
/// of Pantaleo et al. (2010). Estimating `N` betas and `N` residual variances
/// instead of `N(N+1)/2` covariances makes it well conditioned at any window
/// length, at the cost of assuming every correlation between two stocks runs
/// through the single factor.
///
/// See [module-level docs](super::super) for inputs and outputs.
pub fn single_index(config: Config) -> impl VariancePredictor {
    py_segment_module(
        "tradingflow.predictor.variance.single_index",
        config.params(|_| Ok(())),
    )
}
